synapse/storage/user_directory.py

```diff
@@ -660,14 +567,15 @@ def search_user_dir(self, user_id, search_term, limit):
             where_clause = "1=1"
         else:
             join_clause = """
-                LEFT JOIN users_in_public_rooms AS p USING (user_id)
                 LEFT JOIN (
-                    SELECT other_user_id AS user_id FROM users_who_share_rooms
-                    WHERE user_id = ? AND share_private
-                ) AS s USING (user_id)
+                    SELECT other_user_id AS user_id FROM users_who_share_public_rooms
```

Comment:
doesn't this need a where clause? (or the existing where clause being moved to apply to the result of the UNION, assuming the optimizer can handle that)

Reply:
No, we want all users who share any public room, don't we?

Reply:
good point. comments would help.

Reply:
good point, well made.